src/fs/mod.rs

```diff
@@ -118,28 +130,53 @@ impl Filesystem {
         let mut buf: Vec<u8> = vec![0; size];
         let mut total = 0;
 
-        'blocks: for block in file.blocks.iter().skip(chunk_index) {
-            let (_, mut fd) = match cache.get(block).await {
-                Ok(out) => out,
-                Err(_) => {
-                    return Ok(req.reply_error(libc::EIO)?);
-                }
+        'blocks: for block in file_metadata.blocks.iter().skip(chunk_index) {
+            // fhash works as a key inside the LRU
+            let fhash = block.hash.clone();
+            let mut lruf = self.lruf.lock().await;
+
+            // getting the file descriptor from the LRU or from the cache if not found in the LRU
+            let mut fd = if lruf.contains(&fhash) {
+                let file_descriptor = lruf.pop(&fhash).unwrap();
+                drop(lruf);
+                file_descriptor
+            } else {
+                let (_, file_descriptor) = match cache.get(block).await {
+                    Ok(out) => out,
+                    Err(_) => {
+                        return Ok(req.reply_error(libc::EIO)?);
+                    }
+                };
+                file_descriptor
             };
+            
+            // seek to the position <offset>
             fd.seek(SeekFrom::Start(offset as u64)).await?;
 
             loop {
+                // read the file bytes into buf
                 let read = match fd.read(&mut buf[total..]).await {
                     Ok(n) => n,
                     Err(_) => {
                         return Ok(req.reply_error(libc::EIO)?);
                     }
                 };
+
+                // calculate the total size and break if the whole file downloaded
                 total += read;
                 if total >= size {
                     break 'blocks;
                 }
 
+                // read = 0 means that the file not totally read
                 if read == 0 {
+                    
+                    // if this true it means that the last chunk is not totally read
+                    // so we push it inside the LRU
+                    if total%CHUNK_SIZE > 0 {
```

Comment:
I don't believe this check is correct. if read==0 it means you already hit EOF (end of file) so the file has been entirely read anyway. Also total is the total of what you already read so far. 

The only way you can tell if this file has been read to end or not is by checking if the cursor of the file did not reach end of the file yet.
for example
seek + read < file.size
note that a chunk is always CHUNK_SIZE except the last chunk in a file which can be smaller than CHUNK_SIZE.

Reply:
by the way, a cache.get() return (size, File) in case u need to use it.

Reply:
> note that a chunk is always CHUNK_SIZE except the last chunk in a file which can be smaller than CHUNK_SIZE.

You are ignoring what I said the fact that the last chunk in a file can be less that CHUNK_SIZE
If CHUNK_SIZE = 100  and the file size is 250 then there is 3 chunks of sizes (100, 100, and 50) not (100, 100, 100) then the your fomrula will not work.


IMHO you are complicated this more than the necessary, if you wanna know if the file still has data to read. 
- When u do `cache.get()` this already return you the size of the file
- You do seek anyway.
- You know how much you read from this chunk 

So you can simply know if this chunk still has data to be read by doing 
`size > (internal_seek+read_from_chunk)`